closeio/contrib/django/utils.py

```diff
@@ -0,0 +1,69 @@
+import hashlib
+import hmac
+import json
+import logging
+
+from django.conf import settings
+
+logger = logging.getLogger(__name__)
+
+
+def webhook_signature_valid(request):
+    """Verifies the integrity of signed webhooks from Closeio.
+
+    When a webhook subscription is created via the Closeio API you get a secret
+    signature key in the response of the request. This is the key that Closeio
+    uses to sign their webhooks. You can use the key to verify that the request
+    is indeed coming from Closeio (only Closeio should have the signature key)
+    and that the payload was not altered inbetween.
+    This is a simple security mechanism to protect you from false requests to your
+    endpoints that receive the webhooks.
+
+    The signature key for each webhook subscription needs to be stored inside the
+    environment variable `CLOSEIO_WEBHOOK_SIGNATURE_KEYS` as json string containing
+    the webhook subscription ID as key and the signature key as value, e.g.::
+        CLOSEIO_WEBHOOK_SIGNATURE_KEYS = '{"whsub_1": "123", "whsub_2": "456"}'
+
+    More information can be found in the `Closeio Docs`_ in the section "Webhook signatures".
+
+    .. _Closeio Docs: https://developer.close.com/#webhook
+
+    Args:
+        request: an instance of Django's ``HttpRequest`` object
+    """
+    payload = json.loads(request.body)
+    subscription_id = payload.get('subscription_id')
```

Comment:
hmm... I need to think about this one. I think the chance that Closeio will start sending weird stuff like invalid JSON is rather low, but it's possible. You might be right that a 400 response could be better here...

Reply:
After thinking some more about this: We can probably leave it up to the view calling this code to validate the request body. They want to use it after all.

They should only know that validation needs to happen before calling this function. One way to enforce this would be to ask for the subscription id as a function parameter, and only treat the body as a bytes object inside the function, without decoding and reencoding it.